mkdocs_git_authors_plugin/repo.py

```diff
@@ -0,0 +1,774 @@
+    """
+    Results of git blame for a given file.
+
+    Stores a list of tuples with a reference to a
+    Commit object and a list of consecutive lines
+    modified by that commit.
+    """
+
+    def __init__(self, repo: Repo, path: Path):
+        """
+        Instantiante a Page object
+
+        Args:
+            repo: Reference to the global Repo instance
+            path: Absolute path to the page's Markdown file
+        """
+        super().__init__(repo)
+        self._path = path
+        self._sorted = False
+        self._total_lines = 0
+        self._authors = []
+        try:
+            self._execute()
+        except GitCommandError:
+            logging.warning(
+                '%s has not been committed yet. Lines are not counted' % path
+            )
+
+    def add_total_lines(self, cnt: int = 1):
+        """
+        Add line(s) to the count of total lines for the page.
+
+        Arg:
+            cnt: number of lines to add. Default: 1
+        """
+        self._total_lines += cnt
+
+    def authors(self):
+        """
+        Return a sorted list of authors for the page
+
+        The list is sorted once upon first request.
+        Sorting is done by author name.
+
+        Args:
+
+        Returns:
+            sorted list with Author objects
+        """
+        if not self._sorted:
+            repo = self.repo()
+            self._authors = sorted(
+                self._authors,
+                key=repo._sort_key,
+                reverse=repo.config('sort_reverse')
+            )
+            self._sorted = True
+        return self._authors
+
+    def authors_summary(self):
+        """
+        Summarized list of authors to a HTML string
+
+        Args:
+        Returns:
+            str: HTML text with authors
+        """
+
+        authors = self.authors()
+        authors_summary = []
+        for author in authors:
+            contrib = (
+                ' (%s)' % author.contribution(self.path(), str)
+                if self.repo().config('show_contribution')
+                and len(self.authors()) > 1
+                else ''
+            )
+            authors_summary.append(
+                "<a href='mailto:%s'>%s</a>%s" % (
+                    author.email(),
+                    author.name(),
+                    contrib
+                ))
+        authors_summary = ', '.join(authors_summary)
+        return "<span class='git-authors'>%s</span>" % authors_summary
+
+    def _execute(self):
+        """
+        Execute git blame and parse the results.
+        """
+
+        cmd = GitCommand('blame', ['-lts', str(self._path)])
+        cmd.run()
+
+        # Retrieve SHA and content from the line, discarding
+        # file path and line number
+        line_pattern = re.compile('(.*?)\s.*\s*\d\)(\s*.*)')
+
+        for line in cmd.stdout():
+            m = line_pattern.match(line)
+            if m:
+                sha = m.group(1)
+                content = m.group(2).strip()
```

Comment:
> I suggest that we complete polishing this PR and merge it.

Yes, I agree completely

> In the meantime I'll do two things

Great suggestions! I've also opened #17, which I can pick up after we complete this PR

Reply:
I've created https://stackoverflow.com/questions/60523415/robustly-retrieve-sha-and-line-content-with-git-blame-python3 but while finishing it and testing with `git blame --porcelain` I found that we *should* actually use that (assuming that porcelain output can be considered stable), for a specific reason:

The output of the porcelain version gives us all the information we also store in a `Commit` object. that means: by parsing the git-blame output I can pass all relevant information to `Commit`'s initializer, and we don't even need to run `git show` on the commit at all.

I'll give that a try.

Reply:
Good find!

I also found the [git blame implementation from GitPython](https://github.com/gitpython-developers/GitPython/blob/d7b401e0aa9dbb1a7543dde46064b24a5450db19/git/repo/base.py#L781) uses the porcelain output as well. Maybe you can re-use parts of their code for parsing it.

Perhaps it's better to create a separate issue + PR for this? Then we can resolve this discussion. I'd like to merge this PR to master. It will fail the build, but we won't release yet. Then I can work on the unit tests & other issues, building on this PR.

Reply:
Actually went ahead and created https://github.com/timvink/mkdocs-git-authors-plugin/issues/20. Resolving this discussion.

Reply:
I see. Unfortunately I can't assign issues to myself.